### PR TITLE
3820: Group android notifications and properly use the "group summary alert"

### LINKED ISSUE #3820: Doubled notification sound for multiple favorite notifications
Breaking out a smaller step from #3756:

> The "firstOfBatch" logic should be fixed to a "firstOfGroup"; for that a grouping of the single notifications by "group id" should be done before the notify() loop

Currently if there are multiple notifications of one type (e.g. favorite) there will be two notification sounds: One for the first and one for the summary notification for that type.

- [x] I searched or browsed the repo’s other issues to ensure this is not a duplicate.

This can be tested / worked with in a dev version when modifying the two values remoteMarkerId and localMarkerId in NotificationFetcher#fetchNewNotifications`; setting it to a fixed value some hundred below their current value.
That will yield a few notifications at once each time.

### PR DESCRIPTION
**Pull request by Lakoja**
Fixes #3820 

This mainly corrects the "first of batch" logic.
(Each group/batch has a list so the question "only one per this group?" can be answered.)

### CODE REVIEW

File: app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt [hunk 1]
```diff
@@ -65,23 +66,24 @@ class NotificationFetcher @Inject constructor(
                         }
                     }
 
+                    val notificationsGrouped = groupByGroupId(account, notifications)
```
Comment (nikclayton) [inline app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt:69]: Can you use https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/group-by.html here?

File: app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt [hunk 2]
```diff
@@ -11,8 +11,9 @@ import com.keylesspalace.tusky.entity.Marker
 import com.keylesspalace.tusky.entity.Notification
 import com.keylesspalace.tusky.network.MastodonApi
 import com.keylesspalace.tusky.util.isLessThan
-import kotlinx.coroutines.delay
+import java.util.*
```
Comment (nikclayton) [inline app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt:14]: Lint won't like this (wildcard imports)

File: app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt [hunk 3]
```diff
@@ -65,23 +66,24 @@ class NotificationFetcher @Inject constructor(
...
+                    notificationsGrouped.forEach { notificationList ->
+                        notificationList.forEach { notification ->
+                            val androidNotification = NotificationHelper.make(
+                                context,
+                                notificationManager,
+                                notification,
+                                account,
+                                notificationList.size == 1
+                            )
+                            notificationManager.notify(notification.id, account.id.toInt(), androidNotification)
```
Comment (nikclayton) [inline app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt:85]: Why remove the delay? This will result in notifications being dropped.

Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt:85]: See my findings and the documentation link in https://github.com/tuskyapp/Tusky/issues/3756#issuecomment-1605342221

The only thing that drops (grouped) notifications for me is the maximum total number.

File: app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt (hunk 1 above)
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt]: Ok

It's amazing how many helper functions Kotlin has.
Even if it hadn't it would have made sense to write our own extension for this - as this reads much nicer and more concise now...

File: app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt (hunk 2 above)
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt]: Changed

File: app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt (hunk 3 above)
Comment (nikclayton) [reply app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt:83]: Right, that's what confuses me. Because in that comment you note that you saw this message being logged if the delay is removed.

> Package has already posted or enqueued 25 notifications. Not showing more. package=com.keylesspalace.tusky"

So, remove the delay and Android will refuse to show some of the notifications.

Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt:83]: My findings were from my device (and Android 10). With the emulator Logcat shows something like "Silencing noisy Tusky" as an error after 15 or so (group) notifications without delay.

If they are actually "dropped" I need to check again, also on my device.

However dropping we will have in any case for lots of notifications: a) in our code (50) and b) on actual devices (25 on mine).

So maybe we can (must) leave the delay in here.

But we should (must) switch to batched notifications as soon as possible, I think.

Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt:84]: Added delay again. It might help on some devices.

File: app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationHelper.java [hunk 4]
```diff
@@ -710,7 +710,7 @@ public static boolean filterNotification(@NonNull NotificationManager notificati
     }
 
     @Nullable
-    private static String getChannelId(AccountEntity account, Notification notification) {
+    public static String getChannelId(AccountEntity account, Notification notification) {
```
Comment (nikclayton) [inline app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationHelper.java:713]: Is this change needed now?

File: app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt [hunk 5]
```diff
@@ -65,23 +65,29 @@ class NotificationFetcher @Inject constructor(
                         }
                     }
 
+                    val notificationsGrouped = notifications.groupBy { it.type }
```
Comment (nikclayton) [inline app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt:68]:
```suggestion
                    val notificationsByType = notifications.groupBy { it.type }
```

?

File: app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt [hunk 6]
```diff
@@ -65,23 +65,29 @@ class NotificationFetcher @Inject constructor(
...
-                            account,
-                            index == 0
-                        )
-                        notificationManager.notify(notification.id, account.id.toInt(), androidNotification)
-                        // Android will rate limit / drop notifications if they're posted too
-                        // quickly. There is no indication to the user that this happened.
-                        // See https://github.com/tuskyapp/Tusky/pull/3626#discussion_r1192963664
-                        delay(1000.milliseconds)
+
+                    notificationsGrouped.forEach { notificationListEntry ->
```
Comment (nikclayton) [inline app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt:75]:
```suggestion
                    notificationsGrouped.forEach { notificationGroup ->
```

?

File: app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt (hunk 5 above)
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt]: ok

File: app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt (hunk 6 above)
Comment (Lakoja) [reply app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationFetcher.kt]: ok (`notificationsGroup`)

**Review by Tak — APPROVED** [review]
Looks good to me, but let's hold off on merging anything for a bit

**Review by connyduck — APPROVED** [review]

**Comment by Tak** [comment]
@Lakoja If you want to resolve the conflicts, we can merge this for the next release
